### PR TITLE
Expose lapack routines for solving least squares problems

### LINKED ISSUE #195: Expose lapack routines for solving least squares problems
Lapack has the `xGELS` `xGELSY` `xGELSS` and `xGELSD` methods for solving least squares problems (see [here](https://www.netlib.org/lapack/lug/node27.html)), by QR factorization or SVD. It would be nice to expose these to enable direct solving of least squares problems via ndarray-linalg. In fact, I wanted to do that in [this PR](https://github.com/LukeMathWalker/linfa/pull/20) but resolved to solving the linear equations since these methods were not available.

### PR DESCRIPTION
**Pull request by paulkoerbitz**
This PR wraps the lapack routine `xGELSD` with three traits `LeastSquaresSvd`, `LeastSquaresSvdInto` and `LeastSquaresSvdInPlace` similarly to how this is done for other lapack routines in the library.

Fixes #195.

### CODE REVIEW

File: src/least_squares.rs [hunk 1]
```diff
@@ -0,0 +1,780 @@
...
+use ndarray::{s, Array, Array1, Array2, ArrayBase, Axis, Data, DataMut, Ix1, Ix2};
+
+use crate::error::*;
+use crate::lapack::least_squares::*;
+use crate::layout::*;
+use crate::types::*;
+
+pub trait Ix1OrIx2<E: Scalar> {
+    type ScalarOrArray1;
+}
```
Comment (termoshtt) [inline src/least_squares.rs:72]: Is it possible to replace using this trait by [Dimension::Smaller](https://docs.rs/ndarray/0.13.1/ndarray/trait.Dimension.html#associatedtype.Smaller)?
IMU, this is for switching `ScalarOrArray1`, and it seems to be replaced by `Array<E::Real, I::Smaller>`. You can use 0-dimensional array like

```
let mut arr: Array<i32, Ix0> = Array::zeros(());
arr[()] = 1234;
dbg!(arr); // 1234
```

File: src/least_squares.rs [hunk 2]
```diff
@@ -0,0 +1,780 @@
...
+mod tests {
+    use super::*;
+    use approx::AbsDiffEq;
+    use ndarray::{ArcArray1, ArcArray2, Array1, Array2, CowArray};
+    use num_complex::Complex;
+
+    ///////////////////////////////////////////////////////////////////////////
+    /// Test cases taken from the scipy test suite for the scipy lstsq function
+    /// https://github.com/scipy/scipy/blob/v1.4.1/scipy/linalg/tests/test_basic.py
+    ///////////////////////////////////////////////////////////////////////////
```
Comment (termoshtt) [inline src/least_squares.rs:435]:
```suggestion
    //
    // Test cases taken from the scipy test suite for the scipy lstsq function
    // https://github.com/scipy/scipy/blob/v1.4.1/scipy/linalg/tests/test_basic.py
    //

Comment (termoshtt) [reply src/least_squares.rs:435]: Please use `//` style non-doc comment here. `///` means this comment is for the next `#[test]` function.
https://doc.rust-lang.org/reference/comments.html

File: src/least_squares.rs [hunk 3]
```diff
@@ -0,0 +1,780 @@
...
+    /// different array types.
+    ///
+    ///               | least_squares | ls_into | ls_in_place |
+    /// --------------+---------------+---------+-------------+
+    /// Array         | yes           | yes     | yes         |
+    /// ArcArray      | yes           | no      | no          |
+    /// CowArray      | yes           | yes     | yes         |
+    /// ArrayView     | yes           | no      | no          |
+    /// ArrayViewMut  | yes           | no      | yes         |
+    ///////////////////////////////////////////////////////////////////////////
```
Comment (termoshtt) [inline src/least_squares.rs:560]: Please use non-doc comment

File: src/least_squares.rs [hunk 4]
```diff
@@ -0,0 +1,780 @@
...
+        let av = &mut a2.view_mut();
+        let bv = &mut b2.view_mut();
+        let res = av.least_squares_in_place(bv).unwrap();
+        assert_result(&a, &b, &res);
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    /// Test cases taken from the netlib documentation at
+    /// https://www.netlib.org/lapack/lapacke.html#_calling_code_dgels_code
+    ///////////////////////////////////////////////////////////////////////////
```
Comment (termoshtt) [inline src/least_squares.rs:680]: Please use non-doc comment

File: src/least_squares.rs [hunk 5]
```diff
@@ -0,0 +1,780 @@
...
+        let residual_ssq = residual.mapv(|x| x.powi(2)).sum_axis(Axis(0));
+        assert!(result
+            .residual_sum_of_squares
+            .unwrap()
+            .abs_diff_eq(&residual_ssq, 1e-12));
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    /// Testing error cases
+    ///////////////////////////////////////////////////////////////////////////
```
Comment (termoshtt) [inline src/least_squares.rs:743]: Please use non-doc comment

File: src/least_squares.rs (hunk 1 above)
Comment (paulkoerbitz) [reply src/least_squares.rs]: Works like a charm - thank you for this comment. I did look at dimension but didn't understand the `Ix0` case at the time. Nice improvement.

File: src/least_squares.rs (hunk 2 above)
Comment (paulkoerbitz) [reply src/least_squares.rs]: Fixed.

File: src/least_squares.rs (hunk 3 above)
Comment (paulkoerbitz) [reply src/least_squares.rs]: Fixed.

File: src/least_squares.rs (hunk 5 above)
Comment (paulkoerbitz) [reply src/least_squares.rs]: Fixed.

File: src/least_squares.rs (hunk 4 above)
Comment (paulkoerbitz) [reply src/least_squares.rs]: Fixed.

**Comment by paulkoerbitz** [comment]
Thanks for the fast review, that is super encouraging!

I've addressed your comments, would you mind taking another look?

**Comment by termoshtt** [comment]
Thanks!

**Pull request merged**